lib/assert.js

```diff
@@ -302,7 +302,11 @@ function _throws(shouldThrow, block, expected, message) {
     fail(actual, expected, 'Missing expected exception' + message);
   }
 
-  if (!shouldThrow && expectedException(actual, expected)) {
+  if (!shouldThrow &&
+      actual instanceof Error &&
+      typeof message === 'string' &&
+      expectedException(actual, expected) ||
```

Comment:
I'd prefer to see a more liberal use of parens to make it explicit what the evaluation order is that you're using here, mixing `&&` and `||` without parens makes this very difficult to parse. The number of conditions in here adds to that difficulty and it'd be nice to reduce it with some compacting variables—although not essential, just would be nice to make this less terse.